File: plugins/bluenimble-plugin-scripting.javascript/src/main/resources/platform/core/ApiService.js
/**
  Represents a service running in an api under your space <br/>
  <strong>Do not call constructor directly</strong><br/>
  Bluenimble container creates it internally and makes it available for your services to use 	
  @class ApiService
  @classdesc 
*/

/**
  @constructor
  @access private
*/
var ApiService = function (proxy) {

	this.proxy 			= proxy;

	/**	
	  The id of this service 
	  @type {string}
	  @readonly
	*/
	this.id 			= proxy.getId ();

	/**	
	  The name of this service
	  @type {string}
	  @readonly
	*/
	this.name 			= proxy.getName ();
	
	/**	
	  The description of this service
	  @type {string}
	  @readonly
	*/
	this.description 	= proxy.getDescription ();
	
	/**	
	  The endpoint of this service
	  @type {string}
	  @readonly
	*/
	this.endpoint 		= proxy.getEndpoint ();
	
	/**	
	  The security block of this service - if any -
	  @type {JsonObject}
	*/
	this.security 		= proxy.getSecurity ();
	
	/**	
	  The runtime block of this service - if any -
	  @type {JsonObject}
	*/
	this.runtime		= proxy.getRuntime ();
	
	/**	
	  The features block of this service - if any -
	  @type {JsonObject}
	*/
	this.features		= proxy.getFeatures ();
	
	/**	
	  The media block of this service - if any -
	  @type {JsonObject}
	*/
	this.media			= proxy.getMedia ();
	
	/**	
	  The spi block of this service - if any -
	  @type {JsonObject}
	*/
	this.spiDef			= proxy.getSpiDef ();
	
	/**	
	  The error message in case of a failure during deployment of this service
	  @type {string}
	  @readonly
	*/
	this.failure		= proxy.getFailure ();
	
	/**	
	  Get a spec part
	  @type {string}
	  @readonly
	*/
	this.get = function (name) {
		return proxy.getSpecification (name);
	};
	
	/**	
	  The runtime status of this service. Running, Stopped, Paused or Failed
	  @type {string}
	*/
	this.status = function () {
		return proxy.status ().name ();
	};
	/**	
	  Pause this service.<br/>
	  All upcoming requests for this service will be rejected if the service is paused
	*/
	this.pause = function () {
		proxy.pause ();
	};
	/**	
	  Resume this service
	*/
	this.resume = function () {
		proxy.resume ();
	};
	/**	
	  Get a json representation of this service
	  
	  @returns {JsonObject} service as a json object
	*/
	this.toJson = function () {
		return proxy.toJson ();
	};
	
};

File: sdks/bluenimble-jvm-sdk/src/main/java/com/bluenimble/platform/api/ApiService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bluenimble.platform.api;

import com.bluenimble.platform.json.JsonObject;
import com.bluenimble.platform.scripting.Scriptable;

@Scriptable ( name = "ApiService", runtime = "scripting/service")
public interface ApiService extends Manageable {
	
	interface Spec {
		
		String Verb 		= "verb";
		String Endpoint 	= "endpoint";

		String Id 			= "id";
		String Name 		= "name";
		String Description 	= "description";
		
		interface Security		{
			String Enabled		= "enabled";
			String Roles		= "roles";
			String Schemes		= "schemes";
			String Placeholder	= "placeholder";
			String TokenName	= "tokenName";
		}
		
		interface Media	{
			String Resource		= "resource";
			String Engine		= "engine";
			String Processor	= "processor";
			String Extends		= "extends";
			String Cast			= "cast";
		}

		interface Meta	{
			String Tags		= "tags";
			String Links	= "links";
		}
		
		interface Spi		{
			String Function		= "function";
		}

		interface Runtime		{
		}
		
		interface Tracking	{
			String Level		= "level";
			String Feature		= "feature";
		}

		String Status		= "status";
		String Spec 		= "spec";

		String Mock 		= "mock";
		String Output 		= "output";
		String Features 	= "features";
		
	}
	
	ApiVerb 					getVerb 			();
	String 						getEndpoint 		();
	
	String 						getId 				();
	String 						getName 			();
	String 						getDescription 		();
	
	JsonObject 					getSpecification 	();
	
	void 						setHelper			(String key, Object helper);
	Object 						getHelper			(String key);
	
	JsonObject 					getSecurity 		();
	JsonObject 					getRuntime 			();
	JsonObject 					getFeatures 		();
	JsonObject 					getMedia 			();
	JsonObject 					getTracking 		();
	JsonObject 					getSpiDef 			();
	
	ApiStatus					status				();
	void						pause				();
	void						resume				();
	
	ApiServiceSpi				getSpi				();
	
	JsonObject 					getFailure 			();

	JsonObject					toJson 				();

}
